### ISSUE TITLE
[Question] - TF public models and algorithm operation 

### ISSUE THREAD

**Issue by xomolo** [deel-ai/xplique#147]
Hi developers,

I have been testing this library and I have some questions about how it works for Object Detection tasks using TensorFlow models. As far as I am concern, the whole library was built to work with TensorFlow, and you are currently working for PyTorch models’ compatibility. Therefore, I have a few questions about.

_Use case with TensorFlow:_

Downloading a model from TensorFlow public models for Object Detection (obtained from TensorFlow git [1]) it is not accepted by library’s explainers. Doing some type research, I found that those models are not from tf.keras.models module, so I made some modifications to a ModelWrapper to achieve compatibility with non gradient-based algorithms but I finally couldn’t with gradient-based ones. Checking how you have tested these functions; I can see you created an ad hoc model so it can work perfectly.

**My questions are:** Are you working on include explainers’ compatibility for TF official models? Have you noticed about that issue before my post? Will this library work exclusively with handcrafted models? Will be published one model (trained with a public dataset) to test TF code?

_Use case with PyTorch:_

Following your Object Detection tutorial, I have no problems on obtaining explanations. However, there is something that doesn’t make sense for me. It is supposed that model output must this shape for next elements:

_(# Images, # Bounding Boxes per image, Bounding Box coordinates “xyxy” + Prediction’s score + Class logits)_

Instead, you use a ones tensor. As far as I know, PyTorch model’s output doesn’t return raw boxes at first, that’s why (I think) you use ones.

**My questions are:** Does it really matter not doing NMS on models when you fill class logits as ones and still working? In this case, why then logits are needed to do explanation algorithm works?

Here I provide you some code so we can find a solution quicklier
```
import numpy as np
import tensorflow as tf
tf.config.set_visible_devices([], "GPU")
import tensorflow_hub as hub
import xplique
from xplique.attributions import SobolAttributionMethod
from xplique.plots import plot_attributions
import tensorflow as tf
import matplotlib.pyplot as plt
import requests
from PIL import Image

class ModelWrapper():
    def __init__(self, model):
        self.model = model

    def __call__(self, inputs):
        # the model expects uint8 inputs
        inputs = tf.cast(inputs, tf.uint8)
        predictions = self.model(inputs)
        return self.format_predictions(predictions)

    @staticmethod
    def format_predictions(predictions):
        """
        Format predictions from a dictionary to a tuple of tensors.
        This format works with tf-hub models and no gradient method.
        """
        boxes = predictions["raw_detection_boxes"]
        classes_probs = predictions['raw_detection_scores']

        scores = tf.reduce_max(classes_probs[:, :, 1:], axis=-1)
        scores = tf.expand_dims(scores, axis=-1)

        predictions = tf.concat([boxes, scores, classes_probs], axis=-1)
        return predictions
 
# load image
url = "https://unsplash.com/photos/MXvcHk-zCIs/download?force=true&w=640"
image = Image.open(requests.get(url, stream=True).raw).resize((512,512))
tf_image = tf.convert_to_tensor(image)
tf_image = tf_image[tf.newaxis, ...]

# load model from tf-hub
detector = hub.load("https://tfhub.dev/tensorflow/ssd_mobilenet_v2/2")
ode_model_2 = ModelWrapper(detector)
tf_predictions = ode_model_2(tf_image)

# wanted BB from raw detections
tf_prediction = tf_predictions[:, 1525]
"""Rise, Hsic, Sobol"""
explainer = SobolAttributionMethod(ode_model_2, operator=xplique.Tasks.OBJECT_DETECTION, batch_size=1)
explanation = explainer.explain(tf_image, tf_prediction)

# visualize explanation
plot_attributions(explanation, tf_image, img_size=2.,
                  cmap='jet', alpha=0.3, absolute_value=False, clip_percentile=0.5)
plt.show()

# own plotting
plt.subplot(1, 2, 1)
plt.imshow(tf_image[0])
plt.subplot(1, 2, 2)
plt.imshow(explanation[0])
plt.show()
```

Regards,

[1] https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md

**Comment by AntoninPoche** [comment]
Dear @xomolo,

Thank you very much for your interest in our library. The library was indeed built for Tensorflow and now supports PyTorch models through the `xplique.wrappers.TorchWrapper` (for attribution methods). Regarding your questions:

- Are you working on include explainers’ compatibility for TF official models?
    - We do not plan to include native compatibility with TF official models. The diversity of bounding box formats across libraries prevents us from covering them all. Nonetheless, if you wish to contribute and add your wrapper to the list of existing wrappers in the library, you are welcome.
- Have you noticed about that issue before my post?
    - Yes, we knew that most on-the-shelf object detection models are not supported without a wrapper to format outputs.
- Will this library work exclusively with handcrafted models?
    - I am not sure what you mean by handcrafted models, but the API choice we made is that the user should build a wrapper to format model outputs and inputs if needed.
- Will be published one model (trained with a public dataset) to test TF code?
    - I am not sure to understand the question, but we do not plan to host any models. Regarding the tutorials, we did not find easy-to-use TensorFlow models, there always were too many problems with the models, (e.g. not fully differentiable, and not easily downloadable...) However, if we do find easy-to-use TensorFlow models, a corresponding Tutorial will be made.
- Instead, you use a ones tensor. As far as I know, PyTorch model’s output doesn’t return raw boxes at first, that’s why (I think) you use ones.
    - The reason we used ones is because the PyTorch model we used did not predict logits for each bounding box, only the maximum class ID. They do it in the process and return it as the objectness, therefore the information is there but the way the DRise score is computed is not compatible. The DRise score multiplies the position score, the class score, and the objectness score. Therefore, to not influence the final score, we wanted to have always one in the class scores, and setting all values to one has this effect.
- Does it really matter not doing NMS on models when you fill class logits as ones and still working?
    - Removing NMS is necessary for two reasons:
        - First, to explain bounding box predictions in the case of perturbation-based methods, predictions on the perturbed image are matched to the most similar predictions in the original image. The NMS removes most of the predicted bounding boxes, thus, the most similar bounding boxes may be much more distant and it may bias the explanation.
        - Second, predictions before the NMS have a similar number of predicted bounding boxes, thus, they can be treated as tensors and computations can be batched.
- In this case, why then logits are needed to do explanation algorithm works?
    - The DRise score explains the three elements of bounding box predictions, the coordinates, the class of the bounding box, and the objectness. Depending on the model, objectness does not correspond to the class logit and class logits are necessary to explain the class prediction for a given box (as in classification).
- Regarding the code you provided:
    - It appears that in the wrapper call, you cast inputs to `tf.uint8`, to my knowledge, casts are not differentiable, this could explain why you did not manage to make a wrapper that works for gradient-based methods. Indeed, the Xplique library manipulates `tf.float32` which is not compatible.

I hope I answered your questions, and we welcome any other questions you may have.

Best regards,
Antonin Poché

**Comment by cmcastellanos** [comment]
Thanks for your answer, but I still have a question.

Will this library work exclusively with handcrafted models?
I am not sure what you mean by handcrafted models, but the API choice we made is that the user should build a wrapper to format model outputs and inputs if needed.
With handcrafted model I mean an ad hoc model, a model formed by some layers to accomplish libary requirements without any training. At xplique/tests/utils.py > generate_object_detection_model [Line 123 branch: master] I see an adhoc model created with this settings:

```
model = Sequential()
model.add(Input(shape=input_shape))
model.add(Conv2D(4, kernel_size=(2, 2),
                 activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(np.prod(output_shape)))
model.add(Reshape(output_shape))
model.add(Activation('sigmoid'))
model.compile(loss='mae', optimizer='sgd')
```

In this case, is it possible that, if we train an object detector like that one, we can work with object detection explainers for gradient and non-gradient based algorithms?

Regards,

**Comment by xomolo** [comment]
Thanks for your answer, but I still have a question.

+ Will this library work exclusively with handcrafted models?
- I am not sure what you mean by handcrafted models, but the API choice we made is that the user should build a wrapper to format model outputs and inputs if needed.
+ With handcrafted model I mean an ad hoc model, a model formed by some layers to accomplish library requirements without any training. At xplique/tests/utils.py > generate_object_detection_model [Line 123 branch: master] I see an adhoc model created with this settings:

```
model = Sequential()
model.add(Input(shape=input_shape))
model.add(Conv2D(4, kernel_size=(2, 2),
                 activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(np.prod(output_shape)))
model.add(Reshape(output_shape))
model.add(Activation('sigmoid'))
model.compile(loss='mae', optimizer='sgd')
```

In this case, is it possible that, if we train an object detector like that one, we can work with object detection explainers for gradient and non-gradient based algorithms?

Regards,

**Comment by AntoninPoche** [comment]
Dear @xomolo,

When you try to apply Xplique to an object detection model, there are two possibilities:
- Either the model respects the expected output and input formats specified in the [Object Detection Documentation](https://deel-ai.github.io/xplique/latest/api/attributions/object_detection/).
- Or a wrapper is needed to manipulate outputs so that they match the expected format.

Indeed, with object detection, there exist many ways to encode bounding box predictions. Therefore, we had to choose which one to use and force users to adapt their models via wrappers to apply Xplique if their models did not match the specifications.

To answer your question, if the object detection model is differentiable, then building a wrapper with TensorFlow functions should not affect it. Thus you will be able to apply gradient-based and perturbation-based methods. The tutorial is an example of that. Therefore, if you have a TensorFlow model working with float32, you will have no problem using both types of methods.

The code in the `generate_object_detection_model` is aimed at testing the library, it was easier to build a random model and manipulate its outputs than to train and save a model to test afterward.

Best regards,
Antonin Poché
